2/answers.txt

```diff
@@ -0,0 +1,231 @@
+
+The remote in this instance is the location that is a copy of the original branch, whereas the origin is the centralsied point from which the most up to date version of a project would be distributed from.
+
+Q5: What is the benefit of forking a repository and submitting pull requests vs sharing repository access and pushing directly to master?
+
+I believe this is because its far safer for you to have your own copy, with which you can experiment freely with, and risk nothing other than your copy, then when you are satisfied that your work does not break anything or fail.
+You can submit a pull request for the original author to view the changes before integrating them with the original source code. Protecting the project from potential errors.
+
+Sharing repository access and pushing directly to the master could result in unforseen changes, potential for multiple changes from different users to conflict, and other such erroneous behaviour.
+
+Q6: Why are git branches useful? how would you organise a repository to best use branches?
+
+Branches are useful because it allows you to move the head and thus pointer to different instances of the work you are doing, enabling you to seperate out paths of changes, whether thats fixing mulitple seperate issues. Or coming up
+with multiple solutions for the same problem, branching these allows you to pick and choose between them and merge the ones that are useful, and merge changes with the master branch.
+
+To organise them I would simply create a branch for each change you wish to make, roughly speaking, so if you were to make a change to the code to solve a particular bug or functionality requirement then you would title that branch accordingly.
+
+Allowing you to compare and merge those changes individually and check for issues that may arise or errors that are made before combining them into the master branch.
+
+Q7: Why is it important to frequently commit and push your changes to a remote repository (eg your forked repo on GitHub)?
+
+Frequent commit and push will safeguard your work against loss of data if the original is somehow compromised, as well as ensuring that with smaller consistent changes, anything that causes
+a problem elsewhere or breaks something is easily identifiable, committing a lot of work at once increases the chance of a conflict with code that is submitted and will make it harder to
+troubleshoot an error should one arise.
+
+Q8: Why are VM snapshots important?
+
+VM Snapshots when done regularly will allow you to restore a VM to a state much closer to that which it was before an error or failure occurs, saving lots of time and effort in lost work
+Ultimately saving the state of the machine like this is similar to regularly saving your work, if it goes down and you have to restore it, you will have lost less if you snapshot regularly
+restoring it to a state before it experienced a problem.
```

Comment:
VM snapshots are painless as they happen in the background and cause no downtime.

There is an exception to this rule though. Some databases dislike snapshots and will refuse to restore their running state on restore.

For this reason, databases are always backed up using a separate dedicated tool that ensures database integrity, eg mysqldump.